### PR TITLE
feat(cache): registry

### PR DESCRIPTION
**Pull request by cschmatzler**
## Summary

This PR implements a Swift Package Registry on the cache nodes that syncs packages from the Swift Package Index (SPI). This enables Tuist users to resolve Swift dependencies through a read-only registry backed by S3 (`S3_REGISTRY_BUCKET`).

### Key Features

- **Package Sync from Swift Package Index**: An Oban worker (`Cache.Registry.SyncWorker`) runs hourly, fetches the SPI package list, processes a cursor-based batch (limit + allowlist), and enqueues release ingestion jobs while respecting `REGISTRY_SYNC_MIN_INTERVAL_SECONDS` per package
- **Release Ingestion**: `Cache.Registry.ReleaseWorker` lists GitHub tags, downloads source archives (zipball or `git clone` when submodules are detected), extracts `Package.swift` variants, computes checksums, and uploads artifacts + manifests to S3
- **Swift Package Registry API**: Read-only implementation in `CacheWeb.RegistryController`:
  - `GET /api/registry/swift` and `GET /api/registry/swift/availability` - Availability probe
  - `GET /api/registry/swift/identifiers?url=` - Lookup package by repository URL
  - `POST /api/registry/swift/login` - SwiftPM login handshake (no-op)
  - `GET /api/registry/swift/{scope}/{name}` - List package releases
  - `GET /api/registry/swift/{scope}/{name}/{version}` - Get release metadata with checksum
  - `GET /api/registry/swift/{scope}/{name}/{version}.zip` - Download source archive
  - `GET /api/registry/swift/{scope}/{name}/{version}/Package.swift[?swift-version=...]` - Fetch manifest (supports swift-version variants) and emits alternate-manifest `Link` header
- **Metadata Storage**: Package metadata stored as JSON in S3 at `registry/metadata/{scope}/{name}/index.json` with in-memory caching (Cachex) and ETag-based revalidation
- **Distributed Locking**: S3-backed locks coordinate leader sync and reduce duplicate per-package/per-release work across cache nodes
- **Key Normalization**: Consistent lowercase keys and semantic version normalization to match server format

### Architecture

```
SyncWorker (hourly)
    │
    ├── Fetches packages.json from SPI
    ├── Processes a cursor-based batch (limit + allowlist)
    └── Enqueues ReleaseWorker jobs for missing tags
            │
            ├── Downloads GitHub zipball (or clones + zips for submodules)
            ├── Extracts Package.swift variants
            ├── Computes SHA256 checksum
            ├── Uploads to S3_REGISTRY_BUCKET:
            │     - registry/swift/{scope}/{name}/{version}/source_archive.zip
            │     - registry/swift/{scope}/{name}/{version}/Package*.swift
            └── Updates metadata JSON:
                  registry/metadata/{scope}/{name}/index.json

RegistryController
    │
    ├── Reads metadata from S3 (cached, ETag revalidated)
    └── Serves artifacts via X-Accel-Redirect (local when present) or nginx remote proxy to a presigned S3 URL
```

### Configuration

New environment variables:
- `S3_REGISTRY_BUCKET` - Bucket for registry artifacts and metadata
- `REGISTRY_GITHUB_TOKEN` - GitHub token for API access during sync
- `REGISTRY_SYNC_ALLOWLIST` - Optional comma-separated list of repos to sync (supports `org/*`)
- `REGISTRY_SYNC_LIMIT` - Max packages to sync per run (default: 350)
- `REGISTRY_SYNC_MIN_INTERVAL_SECONDS` - Minimum time between syncing the same package (default: 6 hours)

Both limit and interval are extremely conservative to monitor the behavior a bit, especially during initial sync; we also want to avoid impacting the actual cache serving while that sync happens, so load should be kept to a minimum. Will increase this, especially 6 hours -> 15 minutes, once everything is proven stable.

### Cloudflare Registry Load Balancer Setup

<details>
<summary>Setup guide</summary>

# Cloudflare Registry Load Balancer Setup

This guide documents how to configure Cloudflare Load Balancing with geo-steering for the Swift Package Registry at `registry.tuist.dev`.

## Overview

The registry is served by 4 cache nodes distributed globally. Cloudflare Load Balancing routes users to the nearest healthy cache node using geo-steering.

### Cache Node Hostnames

| Region | Hostname |
|--------|----------|
| EU Central | `cache-eu-central.tuist.dev` |
| US East | `cache-us-east.tuist.dev` |
| US West | `cache-us-west.tuist.dev` |
| AP Southeast | `cache-ap-southeast.tuist.dev` |

### 1. Create DNS Record

1. Navigate to **DNS** → **Records** in the Cloudflare dashboard
2. Create a new record:
   - **Type**: CNAME (will be converted to A by Load Balancer)
   - **Name**: `registry`
   - **Target**: Will be managed by Load Balancer
   - **Proxy status**: Proxied (orange cloud)

### 2. Create Origin Pools

Create 4 origin pools, one for each region:

#### Pool: EU Central
- **Name**: `registry-eu-central`
- **Origins**:
  - **Address**: `cache-eu-central.tuist.dev`
  - **Weight**: 1
- **Health Check**: See Step 3

#### Pool: US East
- **Name**: `registry-us-east`
- **Origins**:
  - **Address**: `cache-us-east.tuist.dev`
  - **Weight**: 1
- **Health Check**: See Step 3

#### Pool: US West
- **Name**: `registry-us-west`
- **Origins**:
  - **Address**: `cache-us-west.tuist.dev`
  - **Weight**: 1
- **Health Check**: See Step 3

#### Pool: AP Southeast
- **Name**: `registry-ap-southeast`
- **Origins**:
  - **Address**: `cache-ap-southeast.tuist.dev`
  - **Weight**: 1
- **Health Check**: See Step 3

### 3. Configure Health Checks

Create a health check monitor and attach it to all pools:

- **Name**: `registry-health`
- **Type**: HTTP
- **Path**: `/up`
- **Method**: GET
- **Expected Codes**: 200
- **Interval**: 60 seconds
- **Timeout**: 5 seconds
- **Retries**: 2
- **Follow Redirects**: No

### 4. Create Load Balancer

1. Navigate to **Traffic** → **Load Balancing**
2. Create a new Load Balancer:
   - **Hostname**: `registry.tuist.dev`
   - **Session Affinity**: None (stateless API)
   - **Proxy**: Enabled

### 5. Configure Geo-Steering

Set the traffic steering policy to **Geo Steering** and configure region mappings:

| Cloudflare Region Code | Region Name | Primary Pool | Fallback Order |
|------------------------|-------------|--------------|----------------|
| `WEU` | Western Europe | `registry-eu-central` | US East → US West → AP Southeast |
| `EEU` | Eastern Europe | `registry-eu-central` | US East → US West → AP Southeast |
| `ENAM` | Eastern North America | `registry-us-east` | US West → EU Central → AP Southeast |
| `WNAM` | Western North America | `registry-us-west` | US East → EU Central → AP Southeast |
| `SEAS` | Southeast Asia | `registry-ap-southeast` | US West → EU Central → US East |
| `NEAS` | Northeast Asia | `registry-ap-southeast` | US West → EU Central → US East |
| `OC` | Oceania | `registry-ap-southeast` | US West → EU Central → US East |
| `ME` | Middle East | `registry-eu-central` | AP Southeast → US East → US West |
| `AF` | Africa | `registry-eu-central` | US East → AP Southeast → US West |
| `SAS` | South Asia | `registry-ap-southeast` | EU Central → US West → US East |
| `SAM` | South America | `registry-us-east` | US West → EU Central → AP Southeast |
| `CAM` | Central America | `registry-us-east` | US West → EU Central → AP Southeast |

### 6. Configure Fallback Behavior

Set the default fallback pool order for regions not explicitly mapped:

1. `registry-us-east` (primary fallback)
2. `registry-eu-central`
3. `registry-us-west`
4. `registry-ap-southeast`

### 7. Enable the Load Balancer

1. Review all settings
2. Click **Save and Deploy**
3. Wait for DNS propagation (typically 1-5 minutes)

### Test Health Checks

Verify all pools show as healthy in the Cloudflare dashboard:

1. Navigate to **Traffic** → **Load Balancing** → **Pools**
2. Each pool should show a green "Healthy" status

### Test Geo-Steering

From different geographic locations, verify routing:

```bash
# Check which cache node responds
curl -sI https://registry.tuist.dev/up | grep -i "x-served-by"

# Verify registry availability
curl -s https://registry.tuist.dev/api/registry/swift
# Expected: HTTP 200 OK
```

### Test Failover

1. Temporarily disable one origin pool
2. Verify traffic fails over to the next pool in the fallback order
3. Re-enable the pool and verify it receives traffic again

### Pool Shows Unhealthy

1. Verify the cache node is running: `curl -I https://cache-{region}.tuist.dev/up`
2. Check health check configuration matches the endpoint
3. Review Cloudflare health check logs for error details

### Unexpected Routing

1. Verify geo-steering is enabled (not random or round-robin)
2. Check region mappings are correctly configured
3. Use Cloudflare's "Test" feature to simulate requests from different regions

## References

- [Cloudflare Load Balancing](https://developers.cloudflare.com/load-balancing/)
- [Geo Steering](https://developers.cloudflare.com/load-balancing/understand-basics/traffic-steering/steering-policies/geo-steering/)
- [Health Checks](https://developers.cloudflare.com/load-balancing/monitors/)
- [Region Codes](https://developers.cloudflare.com/load-balancing/reference/region-mapping-api/#list-of-load-balancer-regions)

</details>

### Testing

Comprehensive test coverage for:
- Key normalization edge cases
- Metadata caching and S3 operations
- Registry controller endpoints
- Disk storage operations
- Registry S3 transfer queue enqueuing

### CODE REVIEW

File: cache/lib/cache/registry/sync_worker.ex
Comment (fortmarek) [inline cache/lib/cache/registry/sync_worker.ex:1]: shouldn't this worker replace the sync packages worker in the `server`?

File: cache/lib/cache/registry/github.ex
Comment (fortmarek) [inline cache/lib/cache/registry/github.ex:1]: this is now duplicated with the piece in the `server`. We should either move this to the common library or remove the implementation in the `server` since I don't think it's needed there anymore.

File: cache/lib/cache/registry/sync_worker.ex
Comment (cschmatzler) [reply cache/lib/cache/registry/sync_worker.ex:1]: It needs to be duplicated between both of them for now because
- it's a separate bucket; unlike for server, I separated artifact and registry buckets to not be limited by bucket-based rate limits and generally have some nicer separation
- not storing metadata in a central database means we're storing the metadata as JSON in S3 itself, so the bucket structure is different and we can't just rclone things
- synchronizing packages and getting everything tested will take some time after merging this, so I want to have both side-by-side and only deprecate server once we've ran it a while
- the new registry will be under a different URL (I guess `registry.tuist.dev`), so needs user intervention to replace the old one. Unlike for the caches, we do not control the client for registry, so we cannot do the whole runtime lookup of which node to use, so this will resort to Cloudflare's location-based load balancing instead (see Cloudflare Registry Load Balancer Setup detail in the PR description). I'd give 1 months notice to do so and add a warning in the CLI when running `tuist install`, but this deprecation _and_ the fact that cache-based and server-based registry have separate storage structure means we need to sync in both.

Comment (fortmarek) [reply cache/lib/cache/registry/sync_worker.ex:1]: > the new registry will be under a different URL (I guess registry.tuist.dev), so needs user intervention to replace the old one.

Have you considered doing a redirect? I don't think the `swift` CLI passes through warnings from the server, so we will need to rely on reaching out to users via socials, community forum, etc., so they are aware of the deprecation, which is not ideal :/

Comment (cschmatzler) [reply cache/lib/cache/registry/sync_worker.ex:1]: ohh I didn't know SPM actually follows redirects (because some other Xcode stuff doesn't, so bad assumption on my part)
had the clanker investigate the spec and it does, sooo we can

```
**Cloudflare Redirect Rule:**
- **When**: `(http.host eq "tuist.dev" and starts_with(http.request.uri.path, "/registry/"))`
- **Then**: Dynamic redirect to `concat("https://registry.tuist.dev", substring(http.request.uri.path, 9))`
- **Status**: 301 or 303
```
(that part of the comment is claude-generated :) )

will check this out tomorrow.

Comment (fortmarek) [reply cache/lib/cache/registry/sync_worker.ex:1]: nice!

If that's possible, then we can:
- migrate the registry contents to the new bucket
- remove the old implementation
- merge this with a redirect

Removing the old implementation can be done in a follow-up if you feel that'd be better.

Comment (cschmatzler) [reply cache/lib/cache/registry/sync_worker.ex:1]: yeah I'd merge this as addition, test via new subdomain, then fall over existing path via Cloudflare, then remove the old implementation and bucket space once everything works and doesn't blow up. I'll have a follow-up to remove it from the server

File: cache/lib/cache/registry/github.ex
Comment (cschmatzler) [reply cache/lib/cache/registry/github.ex:1]: addressed in [23cc6ed](https://github.com/tuist/tuist/pull/9086/commits/23cc6ed022f8ca1f7aa008ff3920c5d741d1b8ab)

File: cache/lib/cache/registry/release_worker.ex [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+           "git",
+           [
+             "-c",
+             "url.https://github.com/.insteadOf=git@github.com:",
+             "clone",
+             "--depth",
+             "1",
+             "--branch",
+             tag,
+             clone_url,
```
Comment (fortmarek) [inline cache/lib/cache/registry/release_worker.ex:115]: When the metadata lock is already held, `{:error, :already_locked}` propagates up through `sync_release` → `do_sync_release` → `perform`. Since Oban treats `{:error, reason}` as a job failure, this causes the **entire release worker to retry** — re-downloading the source archive, re-uploading, etc. — even though all artifacts were already uploaded successfully.

The other three lock sites (sync, package, release) handle this gracefully by returning `:ok`, but this one is inconsistent. Consider returning `:ok` here as well (the next sync cycle will pick up the missing metadata), or at least snoozing the job instead of failing it.

File: cache/lib/cache/registry/metadata.ex
Comment (fortmarek) [inline cache/lib/cache/registry/metadata.ex:1]: ohhh, so, the package releases, etc. will no longer be stored in the db but as this JSON file? I don't hate that actually, that's pretty smart 👀

File: cache/lib/cache/registry/sync_cursor.ex [hunk 2]
```diff
@@ -0,0 +1,49 @@
+defmodule Cache.Registry.SyncCursor do
+  @moduledoc false
+
+  alias Cache.Config
+
+  require Logger
+
+  @key "registry/state/sync_cursor.json"
```
Comment (fortmarek) [inline cache/lib/cache/registry/sync_cursor.ex:8]: instead of this file, have you considered adding something like `last_updated_releases_at` in the metadata json we already have for the package?

File: cache/lib/cache/registry/sync_worker.ex [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        packages = apply_allowlist(packages, allowlist)
+
+        case packages do
+          [] ->
+            :ok
+
+          _ ->
+            {batch, next_cursor} = take_batch(packages, limit)
+            Enum.each(batch, &sync_package(&1, token))
+            _ = SyncCursor.put(next_cursor)
```
Comment (fortmarek) [inline cache/lib/cache/registry/sync_worker.ex]: do we want to explicitly ignore the return result?

File: cache/lib/cache/registry/sync_worker.ex [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      handle == pattern
+    end
+  end
+
+  defp registry_sync_allowlist do
+    Application.get_env(:cache, :registry_sync_allowlist)
+  end
+
+  defp registry_sync_limit do
+    Application.get_env(:cache, :registry_sync_limit, @default_limit)
```
Comment (fortmarek) [inline cache/lib/cache/registry/sync_worker.ex:213]: should we consider introducing `environment.ex` for cache as well?

File: cache/lib/cache/disk.ex
Comment (fortmarek) [inline cache/lib/cache/disk.ex:1]: as we add more and more here, I wonder if we should keep only domain-agnostic methods here and then adding a `disk` specific to the registry at `cache/registry/disk.ex`. Same for Xcode and Gradle cache

File: cache/lib/cache/registry/sync_worker.ex (hunk 4 above)
Comment (cschmatzler) [reply cache/lib/cache/registry/sync_worker.ex:213]: Will do in a follow-up to not move more around here

File: cache/lib/cache/disk.ex
Comment (cschmatzler) [reply cache/lib/cache/disk.ex:1]: Yeah we should, will do a refactor in a follow up

File: cache/lib/cache/registry/sync_cursor.ex (hunk 2 above)
Comment (cschmatzler) [reply cache/lib/cache/registry/sync_cursor.ex:8]: Thought about it, but the cursor is a global offset into the full SPI package list, not per-package. It tracks where we left off in the round-robin across all packages, so there's no single package's metadata to store it in. The `updated_at` in each package's metadata is for staleness detection on that specific package, different concern.
